lib/temporal/activity/context.rb

```diff
@@ -1,21 +1,57 @@
+require 'temporal/error/failure'
+
 module Temporal
   class Activity
     class Context
       attr_reader :info
 
-      def initialize(info, heartbeat_proc)
+      def initialize(info, heartbeat_proc, shielded: false)
         @thread = Thread.current
         @info = info
         @heartbeat_proc = heartbeat_proc
+        @cancelled = false
+        @shielded = shielded
+        @mutex = Mutex.new
       end
 
       def heartbeat(*details)
         heartbeat_proc.call(*details)
       end
 
+      def shield(&block)
+        # The whole activity is shielded or called from a nested shield
+        return block.call if @shielded
+
+        mutex.synchronize do
+          @shielded = true
+          result = block.call
+          raise Temporal::Error::CancelledError, 'Unhandled cancellation' if @cancelled
```

Comment:
In theory, you should only throw this if cancelled _since the shield started_. What if I run code that is using `shield` in an `ensure` block after already catching a cancel (with intent to swallow cancel)?

Reply:
Oh, interesting. I haven't thought about this. Is this a legit use-case? Like why would you use a shield if already got a `CancelledError`? I'll still make this change, just wondering if I'm missing something

Reply:
It is a very legit use case. People run cleanup code in finally blocks quite frequently. As for why it would be shielded, they may not know why their finally block is run (they may not know if it was due to cancelled exception or other exception), but they want to make sure that their code can't be cancelled if it wasn't a cancellation.

It is important when "shielding" that people be able to use it side effect free unless a cancellation happens while it is running.

Reply:
That makes sense 👍 I've fixed this and added a test for it — https://github.com/temporalio/sdk-ruby/pull/90/commits/351514269b5d2414e6152aac5b0ab7bc1823575e

Reply:
> What if I run code that is using shield in an ensure block after already catching a cancel (with intent to swallow cancel)?

@cretz Can you give a small python code how that would loook like?